scripts/monitoring/auto_monitor_scale_workers.py

```diff
@@ -0,0 +1,404 @@
+        Namespace="AWS/ECS",
+        MetricName="CPUUtilization",
+        Dimensions=[
+            {"Name": "ClusterName", "Value": cluster_name},
+            {"Name": "ServiceName", "Value": service_name},
+        ],
+        StartTime=start_time,
+        EndTime=end_time,
+        Period=period_seconds,
+        Statistics=["Average", "Maximum", "Minimum"],
+    )
+
+    return response["Datapoints"]
+
+
+def get_memory_metrics_for_challenge(
+    challenge,
+    cluster_name=COMMON_SETTINGS_DICT["CLUSTER"],
+    range_days=1,
+    period_seconds=300,
+):
+    """
+    Get the Memory Utilization of the worker in the challenge.
+    """
+
+    cloudwatch_client = get_boto3_client("cloudwatch", aws_keys)
+
+    start_time = datetime.utcnow() - timedelta(days=range_days)
+    end_time = datetime.utcnow()
+    queue_name = challenge.queue
+    service_name = "{}_service".format(queue_name)
+    response = cloudwatch_client.get_metric_statistics(
+        Namespace="AWS/ECS",
+        MetricName="MemoryUtilization",
+        Dimensions=[
+            {"Name": "ClusterName", "Value": cluster_name},
+            {"Name": "ServiceName", "Value": service_name},
+        ],
+        StartTime=start_time,
+        EndTime=end_time,
+        Period=period_seconds,
+        Statistics=["Average", "Maximum", "Minimum"],
+    )
+
+    return response["Datapoints"]
+
+
+def get_storage_metrics_for_challenge(
+    challenge,
+    cluster_name=COMMON_SETTINGS_DICT["CLUSTER"],
+    range_days=1,
+    period_seconds=300,
+):
+    """
+    Get the Storage Utilization of the worker in the challenge.
+    """
+
+    from datetime import datetime, timedelta
+
+    cloudwatch_client = get_boto3_client("cloudwatch", aws_keys)
+
+    start_time = datetime.utcnow() - timedelta(days=range_days)
+    end_time = datetime.utcnow()
+    queue_name = challenge.queue
+    service_name = "{}_service".format(queue_name)
+
+    response = cloudwatch_client.get_metric_statistics(
+        Namespace="ECS/ContainerInsights",
+        MetricName="EphemeralStorageUtilized",
+        Dimensions=[
+            {"Name": "ClusterName", "Value": cluster_name},
+            {"Name": "ServiceName", "Value": service_name},
+        ],
+        StartTime=start_time,
+        EndTime=end_time,
+        Period=period_seconds,
+        Statistics=["Average"],
+    )
+
+    return response["Datapoints"]
+
+
+# get current CPU and Memory utilization for a challenge
+def current_worker_limit(task_definition_name, metrics):
+    ecs_client = get_boto3_client("ecs", aws_keys)
+    try:
+        response = ecs_client.describe_task_definition(
+            taskDefinition=task_definition_name
+        )
+    except Exception as e:
+        print(f"Error retrieving task definition: {str(e)}")
+        return {}
+
+    task_definition = response.get("taskDefinition", {})
+    return task_definition.get(metrics, 0)
+
+
+def get_new_resource_limit(
```

Comment:
This function adjusts the worker limit based on how much they’re being used:

- **Lower the Limit**: If average utilization is below 25%, the limit is cut in half.
- **Raise the Limit**: If average utilization is above 75%, the limit is doubled.

There’s also room to add more complex logic later if needed.